src/communication/serial_context.hpp

```diff
@@ -0,0 +1,55 @@
+#pragma once
+
+#include <vector>
+
+#include <communication/gathered_vector.hpp>
+#include <spike.hpp>
+
+namespace arb {
+
+struct serial_context {
```

Comment:
If the serial_context is for single machine cases and not doing any communication, why not call it single_machine_context?

Reply:
Can we envisage ever having to distinguish between NUMA domains (as relates to the execution context) in a non-MPI run? If so, we might need to be more careful with the naming. I agree though that `serial_context` is probably not the best name.

Reply:
How about `local_context`? 'Local' in these sense of being an antonym of distributed.

Reply:
I think we should encourage at least one MPI rank per NUMA domain, because getting NUMA right is hard, and we don't have a pressing reason to not use MPI between NUMA regions.

I have gone with `local_context` and `mpi_context`.

These contexts are quite difficult to name... Sam and I spent quite a while discussing whether "context" is the right word. The next step is to implement similar abstractions for threading, GPUs, and so on. These will all be packed into an "execution context", at which point we can enfore more sensible naming.